File: test/tutorial/transaction/transaction.cpp
/* includes; won't compile in tests, in the docs we leave out the comment lines to show them
#include <libdnf5/base/goal.hpp>
#include <libdnf5/repo/package_downloader.hpp>
*/

// Create a goal, which is a class that allows to add items for resolving into
// a transaction.
libdnf5::Goal goal(base);

// Add an RPM package named "one" for installation into the goal.
goal.add_rpm_install("one");

// Resolve the goal, create a transaction object.
//
// The argument is `allow_erasing`, a flag indicating whether to allow removing
// packages in the resolved transaction.
auto transaction = goal.resolve();

// We can iterate over the resolved transaction and inspect the packages.
std::cout << "Resolved transaction:" << std::endl;
for (const auto & tspkg : transaction.get_transaction_packages()) {
    std::cout << tspkg.get_package().get_nevra() << ": " << transaction_item_action_to_string(tspkg.get_action())
              << std::endl;
}

// This class demonstrates user-defined callbacks for the url/package downloads.
//
// The callbacks are implemented by inheriting from the callbacks base class
// and overriding its methods.
//
// We only override one of the callbacks here, see
// `libdnf5::repo::DownloadCallbacks` documentation for a complete list.
class PackageDownloadCallbacks : public libdnf5::repo::DownloadCallbacks {
private:
    int mirror_failure(
        [[maybe_unused]] void * user_cb_data,
        const char * msg,
        [[maybe_unused]] const char * url,
        [[maybe_unused]] const char * metadata) override {
        std::cout << "Mirror failure: " << msg << std::endl;
        return 0;
    }
};

base.set_download_callbacks(std::make_unique<PackageDownloadCallbacks>());

// Download the packages.
transaction.download();

// A class for defining the RPM transaction callbacks.
//
// Again, only a callback for when an RPM package installation starts, for a
// complete list of the callbacks see `libdnf5::rpm::TransactionCallbacks`
// documentation.
class TransactionCallbacks : public libdnf5::rpm::TransactionCallbacks {
    void install_start(const libdnf5::base::TransactionPackage & item, [[maybe_unused]] uint64_t total) override {
        std::cout << transaction_item_action_to_string(item.get_action()) << " " << item.get_package().get_nevra()
                  << std::endl;
    }
};

transaction.set_callbacks(std::make_unique<TransactionCallbacks>());

// Add transaction metadata to be stored in the history database.
transaction.set_description("install package one");

// Run the transaction.
std::cout << std::endl << "Running the transaction:" << std::endl;
transaction.run();

File: test/python3/libdnf5/tutorial/transaction/transaction.py
# Create a goal, which is a class that allows to add items for resolving into
# a transaction.
goal = libdnf5.base.Goal(base)

# Add an RPM package named "one" for installation into the goal.
goal.add_rpm_install("one")

# Resolve the goal, create a transaction object.
#
# The argument is `allow_erasing`, a flag indicating wheter to allow removing
# packages in the resolved transaction.
transaction = goal.resolve()

# We can iterate over the resolved transction and inspect the packages.
print("Resolved transaction:")
for tspkg in transaction.get_transaction_packages():
    print(tspkg.get_package().get_nevra(), ": ",
          libdnf5.base.transaction.transaction_item_action_to_string(tspkg.get_action()))

# This class demonstrates user-defined callbacks for the package downloads.
#
# The callbacks are implemented by inheriting from the callbacks base class
# and overriding its methods.
#
# We only override one of the callbacks here, see
# `libdnf5.repo.DownloadCallbacks` documentation for a complete list.


class PackageDownloadCallbacks(libdnf5.repo.DownloadCallbacks):
    def mirror_failure(self, user_cb_data, msg, url=""):
        print("Mirror failure: ", msg)
        return 0


downloader_callbacks = PackageDownloadCallbacks()
base.set_download_callbacks(
    libdnf5.repo.DownloadCallbacksUniquePtr(downloader_callbacks))

# Download the packages.
transaction.download()

# A class for defining the RPM transaction callbacks.
#
# Again, only a callback for when an RPM package installation starts, for a
# complete list of the callbacks see `libdnf5.rpm.TransactionCallbacks`
# documentation.


class TransactionCallbacks(libdnf5.rpm.TransactionCallbacks):
    def install_start(self, item, total=0):
        print(libdnf5.base.transaction.transaction_item_action_to_string(item.get_action()), " ",
              item.get_package().get_nevra())


transaction_callbacks = TransactionCallbacks()
transaction_callbacks_ptr = libdnf5.rpm.TransactionCallbacksUniquePtr(
    transaction_callbacks)
transaction.set_callbacks(transaction_callbacks_ptr)

# Add transaction metadata to be stored in the history database.
transaction.set_description("install package one")

# Run the transaction.
print("Running the transaction:")
transaction.run()
